BaseLib/ConfigTreeNew.cpp

```diff
@@ -0,0 +1,225 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2015, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "ConfigTreeNew.h"
+
+namespace BaseLib
+{
+
+const char ConfigTreeNew::pathseparator = '/';
+const std::string ConfigTreeNew::key_chars_start = "abcdefghijklmnopqrstuvwxyz";
+const std::string ConfigTreeNew::key_chars = key_chars_start + "_0123456789";
+
+ConfigTreeNew::
+ConfigTreeNew(PTree const& tree,
+              Callback const& error_cb,
+              Callback const& warning_cb)
+    : _tree(&tree), _onerror(error_cb), _onwarning(warning_cb)
+{
+    if (!_onerror) {
+        ERR("ConfigTree: No valid error handler provided.");
+        std::abort();
+    }
+    if (!_onwarning) {
+        ERR("ConfigTree: No valid warning handler provided.");
+        std::abort();
+    }
+}
+
+ConfigTreeNew::
+ConfigTreeNew(PTree const& tree, ConfigTreeNew const& parent,
+              std::string const& root)
+    : _tree(&tree), _path(joinPaths(parent._path, root)),
+      _onerror(parent._onerror), _onwarning(parent._onwarning)
+{
+    checkKeyname(root);
+}
+
+ConfigTreeNew::
+ConfigTreeNew(ConfigTreeNew && other)
+	: _tree(other._tree)
+	, _path(other._path)
+	, _visited_params(std::move(other._visited_params))
+	, _onerror(other._onerror)
+	, _onwarning(other._onwarning)
+{
+	other._tree = nullptr;
+}
+
+ConfigTreeNew::~ConfigTreeNew()
+{
+	if (!_tree) return;
+
+	for (auto const& p : *_tree)
+	{
+		markVisitedDecrement(p.first);
+	}
+
+	for (auto const& p : _visited_params)
+	{
+		if (p.second.count > 0) {
+			warning("Key <" + p.first + "> has been read " + std::to_string(p.second.count)
+					+ " time(s) more than it was present in the configuration tree.");
+		} else if (p.second.count < 0) {
+			warning("Key <" + p.first + "> has been read " + std::to_string(-p.second.count)
+					+ " time(s) less than it was present in the configuration tree.");
+		}
+	}
+}
+
+ConfigTreeNew
+ConfigTreeNew::
+getConfSubtree(std::string const& root)
+{
+    if (auto t = getConfSubtreeOptional(root)) {
+        return std::move(*t);
+    } else {
+        error("Key <" + root + "> has not been found.");
+        return ConfigTreeNew(PTree(), *this, ""); // TODO that will crash
+    }
+}
+
+boost::optional<ConfigTreeNew>
+ConfigTreeNew::
+getConfSubtreeOptional(std::string const& root)
+{
+    checkUnique(root);
+    auto subtree = _tree->get_child_optional(root);
+
+    if (subtree) {
+        markVisited(root);
+        boost::optional<ConfigTreeNew> t;
+        t.emplace(*subtree, *this, root);
+        return std::move(t);
```

Comment:
BaseLib/ConfigTreeNew.cpp:99:16: warning: moving a local object in a return statement prevents copy elision
clang's suggestion:  note: remove std::move call here

@chleh could you explain why making `t` an rvalue is needed?